llm/server.go

```diff
@@ -699,6 +700,20 @@ func (s *llmServer) Completion(ctx context.Context, req CompletionRequest, fn fu
 		}
 	}
 
+	if req.Grammar != "" {
+		if req.Format == "json" {
+			return fmt.Errorf("grammar and format cannot be used together")
+		}
+
+		err := ValidateGrammar(req.Grammar)
```

Comment:
Is it really necessary to validate the grammar? Llama.cpp does that anyway?

Reply:
I showed in my investigation above bad grammars eject the model from memory and makes it reload. The go-side grammar check prevents that.

Reply:
Hm, where is go-side in this PR? I see that you implemented your own validator?

Reply:
Yah, I wrote my own validator in the **grammar.go** file of the commit. I didn't want to obligate this project to some special parsing library, so I tried to just be as straight forward as possible to get some initial validation going. I was a bit paranoid the PR might seem too strange if I did something too esoteric. The suite of tests I think could be useful for whatever go validation evolves into.

I was aiming for something just broadly adequate at validation to protect the lamma.cpp server from the ejections of models. I noticed that  a lot of PRs didn't get accepted because they were pretty simplistic pass throughs.  Talking with someone from the Discord said that some lack of even basic protection over the servers state might have been the reason why. That's sort of why this PR evolved as it did.

Reply:
Oh, I thought "go-side" is some library for grammar checking. :-) Lol.

Reply:
S'all good